### PR TITLE
fix(sql): add support for day of year, week of year and week of month date formatting

### PR DESCRIPTION
**Pull request by Zahlii**
Fixes https://github.com/questdb/questdb/issues/2367

### CODE REVIEW

File: core/src/test/java/io/questdb/std/datetime/millitime/DatesTest.java [hunk 1]
```diff
@@ -127,6 +127,34 @@ public void testDayOfWeek() throws Exception {
         Assert.assertEquals(1, Dates.getDayOfWeekSundayFirst(millis));
     }
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/std/datetime/millitime/DatesTest.java:129]: It would be good to test max posssible values for day, week & week of month ?

File: core/src/main/java/io/questdb/std/datetime/millitime/GenericDateFormat.java [hunk 2]
```diff
@@ -522,9 +535,12 @@ public long parse(CharSequence in, int lo, int hi, DateLocale locale) throws Num
                     pos += Numbers.decodeHighInt(l);
                     break;
```
Comment (bziobrowski) [inline core/src/main/java/io/questdb/std/datetime/millitime/GenericDateFormat.java:537]: Duplicates case blocks ?

File: core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java [hunk 3]
```diff
@@ -143,6 +143,27 @@ public void testFormatDay() throws Exception {
         assertFormat("03", "dd", "2014-04-03T00:00:00.000Z");
     }
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java:145]: to_str and to_date should support the same format so it'd be good to test that parse method supports D w W letters .

File: core/src/test/java/io/questdb/std/datetime/millitime/DatesTest.java (hunk 1 above)
Comment (Zahlii) [reply core/src/test/java/io/questdb/std/datetime/millitime/DatesTest.java:129]: Willa dd some more tests here. I couldn't find any info on BC timestamps though, as none of the tests here seem to contain them - Only pre-unix tests, which i'll add. Also, I am not sure about the "correcT" definition of week of year/week of month. Strictly speaking, this calculation currently just uses days // 7 as the week; however it would probably be more reasonable to have this behave according to the "calendar" week - However, in this case we actually would need something like "sunday first" "monday first" for all of these calcualtions.

File: core/src/main/java/io/questdb/std/datetime/millitime/GenericDateFormat.java (hunk 2 above)
Comment (Zahlii) [reply core/src/main/java/io/questdb/std/datetime/millitime/GenericDateFormat.java:537]: Not sure I can follow here, cases are already grouped, or am I missing something?

File: core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java (hunk 3 above)
Comment (Zahlii) [reply core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java:145]: Isn't this already covered by the three tests below (testFormatDayOfYear/WeekOfYear/WeekOfMonth?

Comment (Zahlii) [reply core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java:161]: Ah, I see what you were referring. I think i added tests for that in last commit.

File: core/src/main/java/io/questdb/std/datetime/millitime/GenericDateFormat.java (hunk 2 above)
Comment (bziobrowski) [reply core/src/main/java/io/questdb/std/datetime/millitime/GenericDateFormat.java:544]: Sorry, my mistake. GH made it look like they were part of same switch statement .

File: core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java [hunk 4]
```diff
@@ -107,6 +107,22 @@ public void testDayOneDigit() throws Exception {
         assertThat("dyyyy", "2014-01-03T00:00:00.000Z", "32014");
     }
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java:109]: Could you add more test cases  for each letter and test min, max values  and add similar tests to TimestampFormatCompilerTest ?

File: core/src/test/java/io/questdb/std/datetime/millitime/DatesTest.java (hunk 1 above)
Comment (Zahlii) [reply core/src/test/java/io/questdb/std/datetime/millitime/DatesTest.java:129]: Test have been extended with BC and corner cases now. Also, the logic seems to  be in-line with postgres to_char according to the documentation.

File: core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java (hunk 4 above)
Comment (Zahlii) [reply core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java:109]: I added more tests to both, please let me know if you think more tests are required

File: core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java [hunk 5]
```diff
@@ -107,6 +107,22 @@ public void testDayOneDigit() throws Exception {
...
 
+    @Test
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java:110]: What's the intent of tests below ? 
They ignore day/week field and return start of month .

File: core/src/test/java/io/questdb/std/datetime/millitime/DatesTest.java (hunk 1 above)
Comment (bziobrowski) [reply core/src/test/java/io/questdb/std/datetime/millitime/DatesTest.java:129]: 👍

File: core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java (hunk 4 above)
Comment (bziobrowski) [reply core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java:109]: Nice job ! Tests look good .

File: core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java (hunk 5 above)
Comment (Zahlii) [reply core/src/test/java/io/questdb/std/datetime/millitime/DateFormatCompilerTest.java:110]: Current behavior when parsing is to ignore the info generated by week of year / week of month as well as day of year, as obviously the week won't allow us to pinpoint a specific day based on the other information, in line with what day of week currently is doing. This simply is there to assert that parsing correctly skips the new flags

**Review by bziobrowski — APPROVED** [review]
